In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/life-expectancy-scaled/life_expectancy_scaled.csv


In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
import statsmodels.api as sm
from sklearn.metrics import mean_squared_error, r2_score

df = pd.read_csv("/kaggle/input/life-expectancy-scaled/life_expectancy_scaled.csv")

X = df.drop(columns=["Life expectancy"])
y = df["Life expectancy"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Linear Regression
X_train_sm = sm.add_constant(X_train)
X_test_sm = sm.add_constant(X_test)
model_lr = sm.OLS(y_train, X_train_sm).fit()
y_pred_lr = model_lr.predict(X_test_sm)

#Random Forest Regressor
model_rf = RandomForestRegressor(random_state=42)
model_rf.fit(X_train, y_train)
y_pred_rf = model_rf.predict(X_test)

print("Linear Regression (statsmodels):")
print(model_lr.summary())

print("\nRandom Forest Regressor:")
print(f"MSE: {mean_squared_error(y_test, y_pred_rf):.4f}")
print(f"R² Score: {r2_score(y_test, y_pred_rf):.4f}")

Linear Regression (statsmodels):
                            OLS Regression Results                            
Dep. Variable:        Life expectancy   R-squared:                       0.957
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     1618.
Date:                Sun, 08 Jun 2025   Prob (F-statistic):               0.00
Time:                        14:01:03   Log-Likelihood:                 355.59
No. Observations:                2350   AIC:                            -645.2
Df Residuals:                    2317   BIC:                            -455.0
Df Model:                          32                                         
Covariance Type:            nonrobust                                         
                                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------

Most influential Predictors : HIV/AIDS , Adult Mortality , Diphtheria(DTP3) , Schooling , Hepatitis B , BMI , Country Encoded , Year .

Although the linear model identifies significant linear predictors, the random forest (which captures non-linear and interaction effects) achieved a higher R² (0.9725 vs 0.957).



In [6]:
from sklearn.metrics import mean_absolute_error
import numpy as np

# Linear Regression (statsmodels) predictions are already in y_pred_lr
# Random Forest predictions are already in y_pred_rf

# ---- Linear Regression Evaluation ----
lr_r2 = r2_score(y_test, y_pred_lr)
lr_mae = mean_absolute_error(y_test, y_pred_lr)
lr_mse = mean_squared_error(y_test, y_pred_lr)
lr_rmse = np.sqrt(lr_mse)

print("\nLinear Regression Performance:")
print(f"R²: {lr_r2:.4f}")
print(f"MAE: {lr_mae:.4f}")
print(f"MSE: {lr_mse:.4f}")
print(f"RMSE: {lr_rmse:.4f}")

# ---- Random Forest
rf_r2 = r2_score(y_test, y_pred_rf)
rf_mae = mean_absolute_error(y_test, y_pred_rf)
rf_mse = mean_squared_error(y_test, y_pred_rf)
rf_rmse = np.sqrt(rf_mse)

print("\nRandom Forest Regressor Performance:")
print(f"R²: {rf_r2:.4f}")
print(f"MAE: {rf_mae:.4f}")
print(f"MSE: {rf_mse:.4f}")
print(f"RMSE: {rf_rmse:.4f}")


Linear Regression Performance:
R²: 0.9623
MAE: 0.1237
MSE: 0.0361
RMSE: 0.1901

Random Forest Regressor Performance:
R²: 0.9725
MAE: 0.0978
MSE: 0.0264
RMSE: 0.1625


R-squared(Proportion of variance explained) : Random Forest Regressor has a greater R-squared.

MAE(Average of absolute errors) : Random Forest Regressor has lower MAE.

MSE(Mean of Squared errors) : Forest Regressor has lower MSE.

RMSE(Root of MSE) : Forest Regressor has lower RMSE.